app/models/html_attachment.rb

```diff
@@ -70,7 +70,7 @@ def url(options = {})
     path_or_url = full_url ? "url" : "path"
     path_helper = "#{type}_html_attachment_#{path_or_url}"
 
-    Whitehall.url_maker.public_send(path_helper, attachable.slug, self, options)
+    Whitehall.url_maker.public_send(path_helper, attachable.slug, content_id, options)
```

Comment:
I tried to find the class definition for UrlMaker to figure out what parameters the constructor takes, but all I can see is [whitehall/url_maker.rb](https://github.com/alphagov/whitehall/blob/main/lib/whitehall/url_maker.rb), which pulls in behaviour from other places. I'm not really sure what the `self` was doing before this change. Dug up a couple of links (https://github.com/alphagov/whitehall/blob/5cba22ae73efabd6e93f8da75b28a2ae88ffdec7/test/support/url_helpers.rb#L13, https://github.com/rails/rails/issues/29992) but still a bit nonplussed).

Could you explain in a comment in the code, what this is doing? E.g. what is the relationship between `attachable.slug` and `content_id`?

Reply:
This bit of the code was particularly horrible, this is what I think is going on (which I'll add in a comment if that seems to be right).

First I had assumed that `public_send` was a method specific to `url_maker` but when it clicked that it was a form of `send` it clarified that it was constructing a method name as a string (`path_helper`) and then calling it. Unfortunately these are all dynamically-generated methods based (I think) on Rails' routing. So Rails is creating methods like `publication_html_attachment_path` with two parameters `publication_id` and `id`, the latter of which I've changed to `content_id` both in the routes file and this method.

However my knowledge of this is somewhat tenuous, and e.g. the tests still passed when I didn't change the routes file even though I _think_ it's necessary, so I'd like to test in integration before going any further.

And the need to support the old URL style still also makes me wonder if the method still needs to _generate_ old URLs for older content still, which will add another level of complexity

Reply:
🤮 Argh yes, I hadn't considered we'd need to maintain two sets of URL-generating methods! I'd look into doing that in the first instance.

If it's a _total_ nightmare, we could think about just getting this PR in a state where it's good to go (with the new URL style only), then work on the rake task in [the next card](https://trello.com/c/3BeDlugo/2761-republish-whitehall-documents-containing-non-english-html-attachments-5), and finally merge & deploy both PRs, and run the rake task. I can't think of a way of doing that without some temporary downtime on all non-English HTML attachments, though, so would be a candidate for out-of-hours work, as well as being high-stakes / big bang release. So I'd certainly feel more comfortable if we could get both styles of URL working simultaneously.

Reply:
I _think_ how it would end up looking is something like this:

```ruby
if locale == "en"
  # keep the old code here
else
  # new code
end
```

But then for the time being the new code should probably actually be something like:

```ruby
if publication_date < switchover_date
  # old code
else
  # new code
end
```

And the routes might take some work to disambiguate too. Since the locale isn't included in the URL, we can't use that but if the format of the content ID and edition ID is guaranteed to be different then we might be able to use a constraint:

```ruby
get "/publications/:publication_id/:content_id" => "_#_", as: "publication_html_attachment", constraints: { content_id: valid_content_id_regex }
get "/publications/:publication_id/:id" => "_#_", as: "publication_html_attachment"
```